File: student_dorm/lib/models/repair_request_model.dart
import 'package:intl/intl.dart';

class RepairRequestModel {
  int requestId = 0;
  String caseCode = "";
  List<String> typeNames;
  String description = "";
  int statusId = 0;
  String statusName = "";
  String createdAt = "";

  RepairRequestModel({
    required this.requestId,
    required this.caseCode,
    required this.typeNames,
    required this.description,
    required this.statusId,
    required this.statusName,
    required this.createdAt,
  });

  DateTime getRepairRequestDateTime() {
    DateFormat inputFormat = DateFormat('dd-MM-yyyy HH:mm:ss');
    return inputFormat.parse(createdAt);
  }

  
  factory RepairRequestModel.fromJson(Map<String, dynamic> json) {
    String rawTypes = json['type_names'] as String? ?? "";
    List<String> parsedTypes = [];
    if (rawTypes.isNotEmpty) {
      parsedTypes = rawTypes.split(',').map((e) => e.trim()).toList();
    }
    // print(json['request_id']);
    // print(json['case_code']);
    // print(parsedTypes);
    // print(json['description']);
    // print(json['progress_status_id']);
    // print(json['progress_status_name']);
    // print(json['report_date']);

    return RepairRequestModel(
      requestId: json['request_id'] as int,
      caseCode: json['case_code'] as String,
      typeNames: parsedTypes,
      description: json['description'] as String? ?? "",
      statusId: json['progress_status_id'] as int? ?? 1,
      statusName: json['progress_status_name'] as String? ?? "รอตรวจสอบ",
      createdAt: json['report_date'] as String,
    );
  }
}


class RepairRequestResponse {
  bool isError = false;
  List<RepairRequestModel> data = [];
  String errorMessage = "";

  RepairRequestResponse({
    required this.isError,
    required this.data,
    required this.errorMessage,
  });

  factory RepairRequestResponse.fromJson(Map<String, dynamic> json) {
    // print(json['data']);
    return RepairRequestResponse(
      isError: json['isError'],
      data: (json['data'] as List)
          .map((item) => RepairRequestModel.fromJson(item))
          .toList(),
      errorMessage: json['errorMessage'],
    );
  }
}

File: modeator_dorm/src/models/repairRequestModel.js
export class RepairRequestModel {
  constructor({
    requestId,
    caseCode,
    studentName,
    contactPhone,
    roomNumber,
    allowEntry,
    typeNames,
    description,
    statusId,
    statusName,
    createdAt,
  }) {
    this.requestId = requestId;
    this.caseCode = caseCode;
    this.studentName = studentName;
    this.contactPhone = contactPhone;
    this.roomNumber = roomNumber;
    this.allowEntry = allowEntry;
    this.typeNames = typeNames;
    this.description = description;
    this.statusId = statusId;
    this.statusName = statusName;
    this.createdAt = createdAt;
  }

  getRepairRequestDate() {
    const [datePart, timePart] = this.createdAt.split(' ');
    const [year, month, day] = datePart.split('-').map(Number);
    const [hour, minute, second] = (timePart || '00:00:00')
      .split(':')
      .map(Number);
    return new Date(year, month - 1, day, hour, minute, second);
  }

  getRepairRequestDateLabel() {
    return this.getRepairRequestDate().toLocaleDateString('th-TH', {
      day: 'numeric',
      month: 'short',
      year: 'numeric',
    });
  }

  static fromJson(json) {
    const rawTypes = json.type_names ?? '';
    const parsedTypes = rawTypes
      ? rawTypes.split(',').map((t) => t.trim())
      : [];

    return new RepairRequestModel({
      requestId: json.request_id,
      caseCode: json.case_code,
      studentName: json.student_name ?? '',
      contactPhone: json.contact_phone ?? '',
      roomNumber: json.room_number ?? '',
      allowEntry: Boolean(json.allow_entry),
      typeNames: parsedTypes,
      description: json.description ?? '',
      statusId: json.progress_status_id ?? 1,
      statusName: json.progress_status_name ?? 'รอตรวจสอบ',
      createdAt: json.report_date,
    });
  }
}

export class RepairRequestResponse {
  constructor({ isError, data, errorMessage }) {
    this.isError = isError;
    this.data = data;               // array ของ RepairRequestModel
    this.errorMessage = errorMessage;
  }

  static fromJson(json) {
    return new RepairRequestResponse({
      isError: json.isError,
      data: (json.data ?? []).map((item) => RepairRequestModel.fromJson(item)),
      errorMessage: json.errorMessage ?? '',
    });
  }
}